optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/variable/index/IndexVariableListener.java

```diff
@@ -48,9 +48,7 @@ public void afterEntityAdded(ScoreDirector<Solution_> scoreDirector, Object enti
 
     @Override
     public void beforeVariableChanged(ScoreDirector<Solution_> scoreDirector, Object entity) {
-        // TODO maybe set all indexes to null
-        // When we support over-constrained planning (e.g. there is not enough Employees to assign all tasks), then
-        // an unassigned planning value should have a null index.
+        retract((InnerScoreDirector<Solution_, ?>) scoreDirector, entity);
```

Comment:
This has a big performance impact but is necessary to avoid undo score corruption during CH.

TODO benchmark this.

Reply:
Disaster.






_ | init-24tasks-8employees | init-50tasks-5employees | init-100tasks-5employees | init-500tasks-20employees
-- | -- | -- | -- | --
Before this PR | 58,635/s | 31,746/s | 21,352/s | 12,865/s
Now | 33,098/s | 10,762/s | 7,035/s | 3,766/s

Reply:
Although I see that these numbers are indeed disastrous, does it matter what they were before, if before was wrong?
How do the numbers look comparing to the current implementation?

Reply:
Updated. The list variable was already slightly worse than chained variable before this PR.

Reply:
Vehicle routing example, main (chained) vs. list variable modifications on top of this branch:


_ | cvrp-n181-k23 | cvrp-n344-k43 | cvrp-n627-k43 | cvrp-n1001-k43
-- | -- | -- | -- | --
main (chained) | 52,377/s | 47,495/s | 37,899/s | 33,887/s
list variable | 27,904/s | 25,516/s | 14,772/s | 10,705/s